docs/Guides/Delay-Accepting-Requests.md

```diff
@@ -0,0 +1,529 @@
+<h1 align="center">Fastify</h1>
+
+# Delay Accepting Requests
+
+## Introduction
+
+Fastify provides several [hooks](../Reference/Hooks.md) useful for a variety of
+situations. One of them is the [`onReady`](../Reference/Hooks.md#onready) hook, which is
+useful for executing tasks *right before* the server starts accepting new
+requests. There isn't, though, a direct mechanism to handle scenarios in which
+you'd like the server to start accepting **specific** requests and denying all
+others, at least up to some point.
+
+Say, for instance, your server needs to authenticate with an OAuth provider in
+order to start serving requests. To do that it'd need to engage in the
+[OAuth Authorization Code Flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow),
+which would require it to listen to two requests from the authentication
+provider:
+
+1. the Authorization Code webhook
+2. the tokens webhook
+
+Until the authorization flow is done you wouldn't be able to serve customer
+requests. What to do then?
+
+There are several solutions for achieving that kind of behavior. Here we'll
+introduce one of such techniques and, hopefully, you'll be able to get things
+rolling asap!
+
+## Solution
+
+### Overview
+
+The proposed solution is one of many possible ways for dealing with the
+proposed scenario and many similar to it. It relies solely on Fastify, so no
+fancy infrastructure tricks or third-party libraries will be necessary.
+
+In order to simplify things we won't be dealing with a precise OAuth flow but,
+instead, simulate a scenario in which some key is needed in order to serve
+a request and that key can only be retrieved in runtime by authenticating with
+an external provider.
+
+The main goal here is to deny requests that would otherwise fail **as early as
+possible** and with some **meaningful context**. That's both useful for the
+server (less resources allocated into a bound-to-fail task) and for the client
+(they get some meaningful information and don't need to wait long for it).
+
+That will be achieved by wrapping into a custom plugin two main features:
+
+1. the mechanism for authenticating with the provider
+[decorating](../Referece/Decorators.md) the `fastify` object with the
+authentication key (`magicKey` from here onwards)
+1. the mechanism for denying requests that would, otherwise, fail
+
+### Hands-on
+
+For this sample solution we'll be using the following tools on a Linux
+environment running the `5.17.1-arch1-1` Kernel version:
+
+- `node.js v16.14.2`
+- `npm 8.5.0`
+- `fastify 3.28.0`
+- `fastify-plugin 3.0.1`
+- `undici 5.0.0`
+
+Say we have the following base server set up at first:
+
+```js
+const Fastify = require('fastify')
+
+const provider = require('./provider')
+
+const server = Fastify({ logger: true })
+const USUAL_WAIT_TIME_MS = 5000
+
+server.get('/ping', (request, reply) => {
```

Comment:
Please do not use arrow functions for handlers throughout this guide. Supply full `function (request, reply) {}` functions.